### ISSUE TITLE
Impure expressions are being propagated

### ISSUE THREAD

**Issue by katrinafyi** [UQ-PAC/aslp#21]
Some impure expressions (e.g. R[0]) are being propagated forward which leads to unsoundness. 

0xa9400400 ldp x0, x1, [x0]
```python
__array _R [ 0 ] = Mem.read.0 {{ 8 }} ( __array _R [ 0 ] [ 0 +: 64 ],8,0 ) ;
__array _R [ 1 ] = Mem.read.0 {{ 8 }} ( 
  add_bits.0 {{ 64 }} ( __array _R [ 0 ] [ 0 +: 64 ],'0000000000000000000000000000000000000000000000000000000000001000' ),8,0 ) ;
```
This should load two words from [x0] and [x0+8], but the second load incorrectly uses the new x0 value. R[0] needs to be read into a constant at the start which is used in both address calculations.

With Transforms.CopyProp disabled:
```python
constant bits ( 64 ) Exp14__5 = __array _R [ 0 ] ;
constant bits ( 64 ) Exp16__5 = Mem.read.0 {{ 8 }} ( Exp14__5 [ 0 +: 64 ],8,0 ) ;
constant bits ( 64 ) Exp18__5 = Mem.read.0 {{ 8 }} ( add_bits.0 {{ 64 }} ( Exp14__5 [ 0 +: 64 ],'0000000000000000000000000000000000000000000000000000000000001000' ),8,0 ) ;
__array _R [ 0 ] = Exp16__5 ;
__array _R [ 1 ] = Exp18__5 ;
```

**Comment by ncough** [comment]
I agree that the duplication of a register access here is undesirable, but I suspect there is a need to clarify the formal correctness of copy-prop and the partial evaluation in general.

Do we consider register/memory loads to be side-effecting operations? Current transforms have been developed with a model where they are't. For instance, copy-prop is currently duplicating loads, as you point out here, and also reordering them where it is "safe". Dead code elimination will remove loads as well, if their results aren't used. Hypothetically CSE could also be coalescing loads, but I'm not sure it currently will.

From a verification perspective, it is easiest to treat the loads as side-effects and preserve their ordering. This is fine for copy-prop and CSE, but I wonder how much additional dead code will no longer be eliminated.

**Comment by katrinafyi** [comment]
I think yes, they should be considered side-effecting. This would simplify the transforms and make it easier to ensure correctness. 

The minor increase in code size is a small compromise to make. I expect most of the unused variable elimination would still succeed, I don't think this affects DCE? Do you have an example? There might be some intermediate variables (like the second code above), but this is ok and simplifying it further would need some kind of def-use analysis. 

How would we go about specifying the correctness of partial evaluation? So far, I've been working on a notion of "identical effects to original code".

**Comment by ncough** [comment]
Across the coverage tests, there are roughly 6k side-effecting loads that are dead. Most are register loads where:
- The destination of the operation is the zero register so the operation is effectively a nop (`adc wzr, w0, w0`)
- One of the operands is a contant/zero register and the result of the operation reduces to a constant (`and x0, xzr, x1`)

This figure is likely biased a little high, due to the how the coverage tests are generated, so leaving them in seems like an acceptable trade-off.

For defining correctness, it would be equivalent side-effect traces between the produced program and the input. We don't have any non-deterministic behaviour that we exploit (I think?), so we don't have to consider something weaker like refinement. Just need to identify the language of side-effects. I imagine something like: state accesses, errors, some interface of functions that aren't inlined.

**Comment by katrinafyi** [comment]
That makes sense. After thinking about it, there are actually two separate but related things here.

It's always incorrect to copy propagate a variable after it has been reassigned, like _R[0] in the first example. I believe this should be prevented by the clobber checks in CopyProp, is that the case? In any case, this would be a bug regardless of if it's side-effecting or not.

Second, are memory/register loads side-effecting? I've realised that the current comparison (specifically in the opt passes) does assume they're _not_ side-effecting; it will remove dead loads. This is actually fine and makes it a bit easier to work with. If that's the assumption carried throughout, it could be left as is, as long as the first copy prop bug is fixed.

**Comment by ncough** [comment]
Ah, I completely missed the actual issue here. Copy-prop didn't support this case, with a TODO kicking the problem down the road. I have a possible fix in #22.

Concerning correctness, I think there are some questions to answer still. Particularly from the perspective of the weak memory work, eliminating reads has some funny effects. This can move to a new issue however.

**Comment by katrinafyi** [comment]
Thanks for the fix, it looks good! I am surprised this wasn't flagged by the coverage testing. It appears the RAM wasn't being compared properly, so I'm working on a small fix for that too.
